# Lab4-Assignment about Named Entity Recognition and Classification

This notebook describes the assignment of Lab 4 of the text mining course. We assume you have succesfully completed Lab1, Lab2 and Lab3 as welll. Especially Lab2 is important for completing this assignment.

**Learning goals**
* going from linguistic input format to representing it in a feature space
* working with pretrained word embeddings
* train a supervised classifier (SVM)
* evaluate a supervised classifier (SVM)
* learn how to interpret the system output and the evaluation results
* be able to propose future improvements based on the observed results


## Credits
This notebook was originally created by [Marten Postma](https://martenpostma.github.io) and [Filip Ilievski](http://ilievski.nl) and adapted by Piek vossen

## [Points: 18] Exercise 1 (NERC): Training and evaluating an SVM using CoNLL-2003

**[4 point] a) Load the CoNLL-2003 training data using the *ConllCorpusReader* and create for both *train.txt* and *test.txt*:**

    [2 points]  -a list of dictionaries representing the features for each training instances, e..g,
    ```
    [
    {'words': 'EU', 'pos': 'NNP'}, 
    {'words': 'rejects', 'pos': 'VBZ'},
    ...
    ]
    ```

    [2 points] -the NERC labels associated with each training instance, e.g.,
    dictionaries, e.g.,
    ```
    [
    'B-ORG', 
    'O',
    ....
    ]
    ```

In [1]:
# %pip install nltk
from nltk.corpus.reader import ConllCorpusReader
### Adapt the path to point to the CONLL2003 folder on your local machine
train = ConllCorpusReader('./CONLL2003/CONLL2003', 'train.txt', ['words', 'pos', 'ignore', 'chunk'])
training_features = []
training_gold_labels = []

for token, pos, ne_label in train.iob_words():
   a_dict = {
      "word.lower()": token.lower(),
      "word.isupper()": token.isupper(),
      "word.istitle()": token.istitle(),
      "word.isdigit()": token.isdigit(),
      "postag": pos,
      "word.shape": ''.join(['X' if c.isupper() else 'X' if c.islower() else 'd' if c.isdigit() else '-' for c in token])
   }
   training_features.append(a_dict)
   training_gold_labels.append(ne_label)  

In [2]:
### Adapt the path to point to the CONLL2003 folder on your local machine
test = ConllCorpusReader('./CONLL2003/CONLL2003', 'test.txt', ['words', 'pos', 'ignore', 'chunk'])

test_features = []
test_gold_labels = []
for token, pos, ne_label in test.iob_words():
    a_dict = {
        "word.lower()": token.lower(),
        "word.isupper()": token.isupper(),
        "word.istitle()": token.istitle(),
        "word.isdigit()": token.isdigit(),
        "postag": pos,
        "word.shape": ''.join(['X' if c.isupper() else 'X' if c.islower() else 'd' if c.isdigit() else '-' for c in token])
    }
    test_features.append(a_dict)
    test_gold_labels.append(ne_label)


**[2 points] b) provide descriptive statistics about the training and test data:**
* How many instances are in train and test?
* Provide a frequency distribution of the NERC labels, i.e., how many times does each NERC label occur?
* Discuss to what extent the training and test data is balanced (equal amount of instances for each NERC label) and to what extent the training and test data differ?

Tip: you can use the following `Counter` functionality to generate frequency list of a list:

In [ ]:
from collections import Counter

# How many instances are in train and test?
num_train_instances = len(training_gold_labels)
num_test_instances = len(test_gold_labels)

print("Number of Instances\n")
print(f"Training:\t{num_train_instances}")
print(f"Test:\t{num_test_instances}")

# Provide a frequency distribution of the NERC labels
train_label_counts = Counter(training_gold_labels)
test_label_counts = Counter(test_gold_labels)

print("\nFrequency Distribution of NERC Labels | Training Data\n")
for label, count in train_label_counts.most_common():
    percentage = (count / num_train_instances) * 100
    print(f"{label}:\t{count} ({percentage:.2f}%)")

print("\nFrequency Distribution of NERC Labels | Test Data\n")
for label, count in test_label_counts.most_common():
    percentage = (count / num_test_instances) * 100
    print(f"{label}:\t{count} ({percentage:.2f}%)")

# To what extent do the training and test data differ?

print("\nDifferences between Training and Test Data:\n")

# Get unique labels from training data
train_unique_labels = set(train_label_counts.keys())
# Get unique labels from test data
test_unique_labels = set(test_label_counts.keys())
# Combine all unique labels from both training and test sets
all_unique_labels_set = train_unique_labels.union(test_unique_labels)
# Sort them for consistent display order
all_labels = sorted(list(all_unique_labels_set))

print(f"{'Label':<2}\t {'Train':<2}\t {'Test':<2}\t {'Diff':<2}\n")
for label in all_labels:
    train_perc = (train_label_counts.get(label, 0) / num_train_instances) * 100
    test_perc = (test_label_counts.get(label, 0) / num_test_instances) * 100
    diff = abs(train_perc - test_perc)
    print(f"{label:<2}\t{train_perc:<2.2f}\t{test_perc:<2.2f}\t{diff:<2.2f}")


Number of Instances

Training:	203621
Test:	46435

Frequency Distribution of NERC Labels | Training Data

O:	169578 (83.28%)
B-LOC:	7140 (3.51%)
B-PER:	6600 (3.24%)
B-ORG:	6321 (3.10%)
I-PER:	4528 (2.22%)
I-ORG:	3704 (1.82%)
B-MISC:	3438 (1.69%)
I-LOC:	1157 (0.57%)
I-MISC:	1155 (0.57%)

Frequency Distribution of NERC Labels | Test Data

O:	38323 (82.53%)
B-LOC:	1668 (3.59%)
B-ORG:	1661 (3.58%)
B-PER:	1617 (3.48%)
I-PER:	1156 (2.49%)
I-ORG:	835 (1.80%)
B-MISC:	702 (1.51%)
I-LOC:	257 (0.55%)
I-MISC:	216 (0.47%)

Differences between Training and Test Data:

Label	 Train	 Test	 Diff

B-LOC	3.51	3.59	0.09
B-MISC	1.69	1.51	0.18
B-ORG	3.10	3.58	0.47
B-PER	3.24	3.48	0.24
I-LOC	0.57	0.55	0.01
I-MISC	0.57	0.47	0.10
I-ORG	1.82	1.80	0.02
I-PER	2.22	2.49	0.27
O 	83.28	82.53	0.75


Both the training and test datasets are highly imbalanced. The $O$ label, makes up the vast majority of instances. But they are similarly imbalanced. Thus the other features are less imbalanced between them both, than individually, in other words, although they differ in scale, the match up in imbalance. This scenario isn't ideal, and much of the classes do still suffer from imbalance between both the training and test sets ($O$, and $B-ORG$ are standouts).

**[2 points] c) Concatenate the train and test features (the list of dictionaries) into one list. Load it using the *DictVectorizer*. Afterwards, split it back to training and test.**

Tip: You’ve concatenated train and test into one list and then you’ve applied the DictVectorizer.
The order of the rows is maintained. You can hence use an index (number of training instances) to split the_array back into train and test. Do NOT use: `
from sklearn.model_selection import train_test_split` here.


In [4]:
# %pip install --upgrade numpy==1.26.4
# %pip install scikit-learn
from sklearn.feature_extraction import DictVectorizer

In [5]:
vec = DictVectorizer()

# Concatenate train and test features
combined_features = training_features + test_features

# Initialize the DictVectorizer
vec = DictVectorizer()

# Fit the vectorizer on the combined features
the_array = vec.fit_transform(combined_features)

# Store the number of training instances to use as the split index
num_train_instances = len(training_features)

# Split the array back into training and test sets using the index
X_train_vectorized = the_array[:num_train_instances]
X_test_vectorized = the_array[num_train_instances:]

**[4 points] d) Train the SVM using the train features and labels and evaluate on the test data. Provide a classification report (sklearn.metrics.classification_report).**

In [6]:
from sklearn import svm

In [7]:
lin_clf = svm.LinearSVC()

In [ ]:
# Train the classifier
from sklearn.metrics import classification_report

lin_clf.fit(X_train_vectorized, training_gold_labels)

# Make predictions on the test data
y_pred = lin_clf.predict(X_test_vectorized)

# Generate the classification report
report = classification_report(test_gold_labels, y_pred)

# Print the classification report
print(report)

c:\dev\anaconda3\envs\PCA25\Lib\site-packages\sklearn\svm\_base.py:1235: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


              precision    recall  f1-score   support

       B-LOC       0.78      0.79      0.79      1668
      B-MISC       0.72      0.69      0.70       702
       B-ORG       0.71      0.57      0.64      1661
       B-PER       0.64      0.58      0.61      1617
       I-LOC       0.64      0.56      0.60       257
      I-MISC       0.63      0.58      0.61       216
       I-ORG       0.66      0.48      0.55       835
       I-PER       0.43      0.62      0.50      1156
           O       0.98      0.99      0.99     38323

    accuracy                           0.93     46435
   macro avg       0.69      0.65      0.66     46435
weighted avg       0.93      0.93      0.92     46435



In [13]:
# For analysis

from collections import Counter

# Count occurrences of each NER label
label_counts = Counter(training_gold_labels)

all_labels = ['O', 'B-PER', 'I-PER', 'B-ORG', 'I-ORG', 'B-LOC', 'I-LOC', 'B-MISC', 'I-MISC']
items = []
for label in all_labels:
    items.append((label, label_counts.get(label, 0)))

items.sort(key=lambda x: x[1])

for i in items:
    print(f"{i[0]}:\t {i[1]}")

I-MISC:	 1155
I-LOC:	 1157
B-MISC:	 3438
I-ORG:	 3704
I-PER:	 4528
B-ORG:	 6321
B-PER:	 6600
B-LOC:	 7140
O:	 169578


The classifier performs well on the $O$ label because it is the most frequent label. That also makes sense because most words don't correspond to anything [source](https://huggingface.co/learn/llm-course/en/chapter7/2#:~:text=O%20means%20the%20word%20doesn%E2%80%99t%20correspond%20to%20any%20entity.).
Ignoring the $O$ case (which is what NERC is really about) the $B-LOC$ feature performs the best. This makes sense as it is the 2nd most frequently occuring instance, and because a $B-LOC$ is usually capitalized (or has some other distinct characteristic) they are more recognizable.

On the contrary the classifier performs poorly on less frequent labels like $I-LOC$, $I-MISC$.

All other features have a similar performance pattern, consistent with the number of instances.

However, this is not the full picture. LSVMs aren't sequential models, which works poorly in most instances because a tag like $I-PER$ depends heavily on its context. A lack of context in combination with class imbalance is what causes a lot of performance loss in this situation.

Overall the model performs quite accurately at a weighted average of $92\%$.

**[6 points] e) Train a model that uses the embeddings of these words as inputs. Test again on the same data as in 2d. Generate a classification report and compare the results with the classifier you built in 2d.**

In [10]:
# %pip install gensim
import numpy as np
from gensim.downloader import load
from sklearn.calibration import LinearSVC 

# load in a pretrained embedding model gensim
glove_vectors = load("glove-wiki-gigaword-100")

# Load training and test data from CONLL2003 corpus
train_data = ConllCorpusReader('./CONLL2003/CONLL2003', 'train.txt', ['words', 'pos', 'ignore', 'chunk'])
test_data = ConllCorpusReader('./CONLL2003/CONLL2003', 'test.txt', ['words', 'pos', 'ignore', 'chunk'])

# A function to get the embedding of a certain word.
def get_embed (word):
    word = word.lower()
    if word in glove_vectors:
        return glove_vectors[word]
    else:
        return np.zeros(glove_vectors.vector_size)
   

# Estabsihing the training data 
X_train = []
y_train = []
X_test  = []
y_test =  []

for sentence in train_data.iob_sents():
    for word, pos, label in sentence:
        X_train.append(get_embed(word))
        y_train.append(label)

#  Estabsihing the testing data 
for sentence in test_data.iob_sents():
    for word, pos, label in sentence:
        X_test.append(get_embed(word))
        y_test.append(label)
       


X_train = np.array(X_train)
X_test = np.array(X_test)


clf = LinearSVC(max_iter=5000)
clf.fit(X_train, y_train)


# Make predictions on the test data
y_pred = clf.predict(X_test)

# Generate the classification report
report = classification_report(y_test, y_pred)

# Print the classification report
print(report)

c:\dev\anaconda3\envs\PCA25\Lib\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
c:\dev\anaconda3\envs\PCA25\Lib\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


              precision    recall  f1-score   support

       B-LOC       0.66      0.73      0.69      1668
      B-MISC       0.76      0.54      0.63       702
       B-ORG       0.63      0.46      0.53      1661
       B-PER       0.70      0.60      0.65      1617
       I-LOC       0.00      0.00      0.00       257
      I-MISC       0.00      0.00      0.00       216
       I-ORG       0.50      0.02      0.03       835
       I-PER       0.60      0.42      0.49      1156
           O       0.93      0.99      0.96     38323

    accuracy                           0.90     46435
   macro avg       0.53      0.42      0.44     46435
weighted avg       0.88      0.90      0.88     46435



c:\dev\anaconda3\envs\PCA25\Lib\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


### Differences between the two reports:

For the overall preformance: 
Accuracy: 1d (93%). which is close to 1e (92%). For the Marco f1 1d again outprefromed 1e, (0,66) > (0,56) the extra features in 1d provides a better view across rare classes. Finally, regarding weighted F1: 1d(0.93) has roughly a similar value as 1e(0.91). If we take a specific look at the enties, we see that there is a major significants in less frequent enties (e.g.I-LOC, I-MISC, I-ORG). 1d (F1: 0.55-0.61) strongly outprefoms 1e (F1: 0.00-0.43), this is due to the fact that the model build in 1e failed to address rare patterns through the lack of explicit featers (e.g. POS, word shapes). On the other side, we see a less major difference between B-LOC and B-PER. The recall for 1e is higher for B-LOC with 0.83 compared to 0.79, probably has 1e a better detection for particiale enteties regarding lower case words. 1d has a higher recall on precision fro B-PER with 0.72 against 0.64. The extra features decreases the change of false postives calls.

The key reasons for distinction for the two models are: extra featers, struggels with embedding and class imbalance, because there is none. 1d has more featers regarding POS word shapes and capitalizaiton, while 1e only focuses on lower case words. 1e struggles with handleling zero vectors and for example for I-LOC class imblancing occurs due to suffering from class weighting. 

1e has did a great try on some entities, but lacks the finess to outperform the overall preformance of 1d.

## [Points: 10] Exercise 2 (NERC): feature inspection using the [Annotated Corpus for Named Entity Recognition](https://www.kaggle.com/abhinavwalia95/entity-annotated-corpus)
**[6 points] a. Perform the same steps as in the previous exercise. Make sure you end up for both the training part (*df_train*) and the test part (*df_test*) with:**
* the features representation using **DictVectorizer**
* the NERC labels in a list

Please note that this is the same setup as in the previous exercise:
* load both train and test using:
    * list of dictionaries for features
    * list of NERC labels
* combine train and test features in a list and represent them using one hot encoding
* train using the training features and NERC labels

In [11]:
import pandas as pd
from sklearn.feature_extraction import DictVectorizer
from sklearn.svm import LinearSVC
from sklearn.metrics import classification_report
from sklearn.preprocessing import LabelEncoder

df = pd.read_csv('ner_dataset.csv', encoding='latin1', on_bad_lines='skip')
df = df.fillna(method='ffill')
df = df[['Word', 'POS', 'Tag']]

# Split dataset
df_train = df[:100000]
df_test = df[100000:120000]

def build_features(dataframe):
    features = []
    labels = []
    for index, row in dataframe.iterrows():
        word = row['Word']
        pos = row['POS']
        tag = row['Tag']
        feature = {
            'word.lower()': word.lower(),
            'word.isupper()': word.isupper(),
            'word.istitle()': word.istitle(),
            'word.isdigit()': word.isdigit(),
            'postag': pos,
            'word.shape': ''.join(['X' if c.isupper() else 'x' if c.islower() else 'd' if c.isdigit() else '-' for c in word])
        }
        features.append(feature)
        labels.append(tag)
    return features, labels

train_features, train_labels = build_features(df_train)
test_features, test_labels = build_features(df_test)

vec = DictVectorizer()
combined_features = train_features + test_features
X_all = vec.fit_transform(combined_features)

X_train_vec = X_all[:len(train_features)]
X_test_vec = X_all[len(train_features):]


C:\Users\jalen\AppData\Local\Temp\ipykernel_24588\1523181299.py:8: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df = df.fillna(method='ffill')


**[4 points] b. Train and evaluate the model and provide the classification report:**
* use the SVM to predict NERC labels on the test data
* evaluate the performance of the SVM on the test data

Analyze the performance per NERC label.

In [12]:
le = LabelEncoder()
y_train_enc = le.fit_transform(train_labels)
y_test_enc = le.transform(test_labels)

clf = LinearSVC(max_iter=1000)
clf.fit(X_train_vec, y_train_enc)

y_pred_enc = clf.predict(X_test_vec)
y_pred = le.inverse_transform(y_pred_enc)
y_test_labels = le.inverse_transform(y_test_enc)

print("report\n")
print(classification_report(y_test_labels, y_pred))

report

              precision    recall  f1-score   support

       B-art       1.00      0.50      0.67         4
       B-eve       0.00      0.00      0.00         0
       B-geo       0.78      0.77      0.77       741
       B-gpe       0.97      0.92      0.95       296
       B-nat       1.00      0.50      0.67         8
       B-org       0.67      0.57      0.62       397
       B-per       0.72      0.59      0.65       333
       B-tim       0.91      0.76      0.83       393
       I-art       0.00      0.00      0.00         0
       I-eve       0.00      0.00      0.00         0
       I-geo       0.73      0.49      0.59       156
       I-gpe       1.00      0.50      0.67         2
       I-nat       0.80      1.00      0.89         4
       I-org       0.58      0.45      0.51       321
       I-per       0.47      0.80      0.59       319
       I-tim       0.41      0.08      0.14       108
           O       0.98      0.99      0.99     16918

    accuracy      

c:\dev\anaconda3\envs\PCA25\Lib\site-packages\sklearn\svm\_base.py:1235: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(
c:\dev\anaconda3\envs\PCA25\Lib\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
c:\dev\anaconda3\envs\PCA25\Lib\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
c:\dev\anaconda3\envs\PCA25\Lib\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero


An overall 94% accuracy was achieved by the SVM model. However, the performance varied a lot across different NERC labels. Where it performed extremely well was in the ‘O’, “non-entity” class. The precision was 0.98 and the recall was 0.99. This was likely because of the large number of non-entity tokens in the dataset. Strong performances were also shown for the geo-political entities, in short B-gpe, and time expressions, B-tim. The f1-scores were 0.95 and 0.83.
The SVM did struggle with entities that were less frequent like B-art, B-eve, I-art and B-nat. The f1-scores for those were either 0.00 or very low. This was because of limited examples and poor recognition. Another thing that showed a lower recall were Inside tags, I-labels. Especially I-geo and I-tim. This suggests that there was difficulty with the recognition of multi-token or nested entities. These imbalances are highlighted by the macro average F1-score of 0.56. On the contrary, the weighted average F1-score, 0.94, is lopsided by the strong performance in the ‘O’ class.


## End of this notebook